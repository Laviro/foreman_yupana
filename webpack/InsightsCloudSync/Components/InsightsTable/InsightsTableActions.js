import URI from 'urijs';
import { push } from 'connected-react-router';
import { get } from 'foremanReact/redux/API';
import { selectQueryParams } from './InsightsTableSelectors';
import { INSIGHTS_PATH } from '../../InsightsCloudSyncConstants';
import {
  columns,
  INSIGHTS_HITS_API_KEY,
  INSIGHTS_HITS_PATH,
  INSIGHTS_SET_SELECTED_IDS,
  INSIGHTS_HIDE_SELECT_ALL_ALERT,
} from './InsightsTableConstants';

export const fetchInsights = (queryParams = {}) => (dispatch, getState) => {
  const { page, perPage, query, sortBy, sortOrder } = {
    ...selectQueryParams(getState()),
    ...queryParams,
  };

  dispatch(
    get({
      key: INSIGHTS_HITS_API_KEY,
      url: INSIGHTS_HITS_PATH,
      params: {
        page,
        per_page: perPage,
        search: query,
        order: `${sortBy} ${sortOrder}`,
      },
    })
  );

  const uri = new URI();
  uri.search({
    page,
    per_page: perPage,
    search: query,
    sort_by: sortBy,
    sort_order: sortOrder,
  });

  dispatch(
    push({
      pathname: INSIGHTS_PATH,
      search: uri.search(),
    })
  );

  dispatch(hideSelectAllAlert());
};

export const hideSelectAllAlert = () => ({
  type: INSIGHTS_HIDE_SELECT_ALL_ALERT,
});

export const selectByIds = (selectedIds = {}, showSelectAllAlert = false) => ({
  type: INSIGHTS_SET_SELECTED_IDS,
  payload: { selectedIds, showSelectAllAlert },
});

export const selectAll = () => (dispatch, getState) => {
  const { query } = { ...selectQueryParams(getState()) };

  const handleSuccess = response => {
    const selectedIds = {};
    response.data.hits.forEach(({ id }) => {
      selectedIds[id] = true;
    });
    dispatch(selectByIds(selectedIds, true));
  };
  dispatch(
    get({
      key: `${INSIGHTS_HITS_API_KEY}_ALL`,
      url: INSIGHTS_HITS_PATH,
      params: {
        search: query,
      },
      handleSuccess,
    })
  );
};

export const clearAllSelection = selectByIds; // passing nothing to `selectByIds` will clear all selection

export const onTableSort = (_event, index, direction) => {
  // The checkbox column shifts the data columns by 1;
  const { sortKey } = columns[index - 1];
  return fetchInsights({
    sortBy: sortKey,
    sortOrder: direction,
    page: 1,
  });
};

export const onTableSetPage = (_event, pageNumber) =>
  fetchInsights({ page: pageNumber });

export const onTablePerPageSelect = (_event, perPageNumber) =>
  fetchInsights({ perPage: perPageNumber });

export const onTableSelect = (
  _event,
  isSelected,
  rowId,
  results,
  prevSelectedIds
) => dispatch => {
  const selectedIds = { ...prevSelectedIds };
  let showSelectAllAlert = false;
  // for select all
  if (rowId === -1) {
    if (!isSelected) return dispatch(clearAllSelection());

    results.forEach(row => {
      selectedIds[row.id] = true;
    });

    showSelectAllAlert = true;
  } else {
    isSelected
      ? (selectedIds[results[rowId].id] = true)
      : delete selectedIds[results[rowId].id];
  }

  return dispatch(selectByIds(selectedIds, showSelectAllAlert));
};
