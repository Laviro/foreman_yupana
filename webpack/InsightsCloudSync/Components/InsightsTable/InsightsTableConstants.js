/* eslint-disable camelcase */
import React from 'react';
import {
  InsightsLabel,
  Section,
} from '@redhat-cloud-services/frontend-components';
import { sortable, cellWidth } from '@patternfly/react-table';
import { CheckCircleIcon } from '@patternfly/react-icons';
import { translate as __ } from 'foremanReact/common/I18n';
import { foremanUrl } from '../../../ForemanRhCloudHelpers';

export const totalRiskFormatter = ({ title: totalRisk }) => ({
  children: (
    <Section className="insights-total-risk" type="icon-group">
      <InsightsLabel value={totalRisk} />
    </Section>
  ),
});

export const hasPlaybookFormatter = ({ title: hasPlaybook }) => ({
  children: hasPlaybook ? <CheckCircleIcon color="green" /> : __('No'),
});

export const columns = [
  {
    sortKey: 'hostname',
    title: __('Hostname'),
    transforms: [cellWidth(20), sortable],
  },
  {
    sortKey: 'title',
    title: __('Recommendation'),
    transforms: [cellWidth(50), sortable],
  },
  {
    sortKey: 'total_risk',
    title: __('Total Risk'),
    transforms: [cellWidth(15), sortable],
    cellTransforms: [totalRiskFormatter],
  },
  {
    title: __('Playbook'),
    transforms: [cellWidth(15)],
    cellTransforms: [hasPlaybookFormatter],
  },
];

export const paginationTitles = {
  items: __('items'),
  page: __('page'),
  itemsPerPage: __('Items per page'),
  perPageSuffix: __('per page'),
  toFirstPage: __('Go to first page'),
  toPreviousPage: __('Go to previous page'),
  toLastPage: __('Go to last page'),
  toNextPage: __('Go to next page'),
  optionsToggle: __('Items per page'),
  currPage: __('Current page'),
  paginationTitle: __('Pagination'),
};

export const INSIGHTS_HITS_PATH = foremanUrl('/insights_cloud/hits');

export const INSIGHTS_HITS_API_KEY = 'INSIGHTS_HITS';

export const INSIGHTS_SET_SELECTED_IDS = 'INSIGHTS_SET_SELECTED_IDS';

export const INSIGHTS_SET_SELECT_ALL_ALERT = 'INSIGHTS_SET_SELECT_ALL_ALERT';

export const INSIGHTS_SET_SELECT_ALL = 'INSIGHTS_SET_SELECT_ALL';
