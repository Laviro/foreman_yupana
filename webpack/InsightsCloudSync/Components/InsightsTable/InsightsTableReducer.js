import Immutable from 'seamless-immutable';
import {
  INSIGHTS_SET_SELECTED_IDS,
  INSIGHTS_HIDE_SELECT_ALL_ALERT,
} from './InsightsTableConstants';

const initialState = Immutable({
  selectedIds: {},
  showSelectAllAlert: false,
});

export default (state = initialState, { payload, type }) => {
  switch (type) {
    case INSIGHTS_SET_SELECTED_IDS:
      return state.merge(payload);
    case INSIGHTS_HIDE_SELECT_ALL_ALERT:
      return state.merge({ showSelectAllAlert: false });
    default:
      return state;
  }
};
