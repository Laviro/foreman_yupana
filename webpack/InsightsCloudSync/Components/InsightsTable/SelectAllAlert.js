import React from 'react';
import PropTypes from 'prop-types';
import { Alert, AlertActionLink } from '@patternfly/react-core';
import { translate as __, sprintf } from 'foremanReact/common/I18n';

const SelectAllAlert = ({
  selectedIds,
  itemCount,
  showSelectAllAlert,
  selectAll,
  clearAllSelection,
  isAllSelected,
}) => {
  if (!showSelectAllAlert) return null;
  const selectedCount = Object.keys(selectedIds).length;
  if (!selectedCount) return null;
  if (!isAllSelected && selectedCount !== itemCount) {
    return (
      <Alert
        isInline
        variant="info"
        title={sprintf(
          'All %s recommendations on this page are selected.',
          selectedCount
        )}
        actionLinks={
          <AlertActionLink onClick={selectAll}>
            {sprintf('Select all %s recommendations', itemCount)}
          </AlertActionLink>
        }
      />
    );
  }

  return (
    <Alert
      isInline
      variant="info"
      title={sprintf('All %s recommendations are selected.', itemCount)}
      actionLinks={
        <React.Fragment>
          <AlertActionLink onClick={clearAllSelection}>
            {__('Clear Selection')}
          </AlertActionLink>
        </React.Fragment>
      }
    />
  );
};

SelectAllAlert.propTypes = {
  selectedIds: PropTypes.object,
  itemCount: PropTypes.number,
  showSelectAllAlert: PropTypes.bool,
  selectAll: PropTypes.func.isRequired,
  clearAllSelection: PropTypes.func.isRequired,
  isAllSelected: PropTypes.bool,
};

SelectAllAlert.defaultProps = {
  selectedIds: {},
  itemCount: 0,
  showSelectAllAlert: false,
  isAllSelected: false,
};

export default SelectAllAlert;
