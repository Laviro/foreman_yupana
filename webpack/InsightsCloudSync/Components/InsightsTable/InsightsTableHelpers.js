/* eslint-disable camelcase */
import { columns } from './InsightsTableConstants';

export const modifySelectedRows = (hits, selectedIds, showSelectAllAlert) => {
  if (hits.length === 0) return [];

  return hits
    .asMutable()
    .map(({ id, hostname, title, total_risk, has_playbook }) => {
      const disableCheckbox = !has_playbook;
      return {
        cells: [hostname, title, total_risk, has_playbook],
        disableCheckbox,
        id,
        /** The main table checkbox will be seen as selected only if all rows are selected,
         * in this case we need to select also the disabled once and hide it with css */
        selected: selectedIds[id] || (disableCheckbox && showSelectAllAlert),
      };
    });
};

export const getSortColumnIndex = sortBy => {
  let colIndex = 0;
  columns.forEach((col, index) => {
    if (col.sortKey === sortBy) {
      // The checkbox column shifts the data columns by 1;
      colIndex = index + 1;
    }
  });
  return colIndex;
};

export const getPerPageOptions = (urlPerPage, appPerPage) => {
  const initialValues = new Set([5, 10, 15, 25, 50]);
  initialValues.add(appPerPage);
  urlPerPage && initialValues.add(urlPerPage);
  const options = [...initialValues].sort((a, b) => a - b);
  return options.map(value => ({ title: value.toString(), value }));
};
