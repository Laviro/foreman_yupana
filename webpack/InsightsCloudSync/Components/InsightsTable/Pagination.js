import React from 'react';
import PropTypes from 'prop-types';
import { useDispatch, useSelector } from 'react-redux';
import {
  Pagination as PfPagination,
  PaginationVariant,
} from '@patternfly/react-core';
import { useForemanSettings } from 'foremanReact/Root/Context/ForemanContext';
import { onTablePerPageSelect, onTableSetPage } from './InsightsTableActions';
import { getPerPageOptions } from './InsightsTableHelpers';
import {
  selectItemCount,
  selectPage,
  selectPerPage,
} from './InsightsTableSelectors';

const Pagination = ({ variant }) => {
  const dispatch = useDispatch();
  const onSetPage = (e, pageNumber) => dispatch(onTableSetPage(e, pageNumber));
  const onPerPageSelect = (e, perPageNumber) =>
    dispatch(onTablePerPageSelect(e, perPageNumber));
  const itemCount = useSelector(state => selectItemCount(state));
  const urlPerPage = useSelector(state => selectPerPage(state));
  const page = useSelector(state => selectPage(state));
  const { perPage: appPerPage } = useForemanSettings();
  const perPage = urlPerPage || appPerPage;

  return (
    <PfPagination
      itemCount={itemCount}
      widgetId={`recommendation-pagination-${variant}`}
      perPage={perPage}
      page={page}
      variant={PaginationVariant[variant]}
      onSetPage={onSetPage}
      onPerPageSelect={onPerPageSelect}
      perPageOptions={getPerPageOptions(urlPerPage, appPerPage)}
    />
  );
};

Pagination.propTypes = {
  variant: PropTypes.string,
};

Pagination.defaultProps = {
  variant: 'top',
};

export default Pagination;
