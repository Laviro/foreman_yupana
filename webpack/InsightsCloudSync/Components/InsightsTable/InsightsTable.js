/* eslint-disable react-hooks/exhaustive-deps */
import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { Table, TableHeader, TableBody } from '@patternfly/react-table';
import { useForemanSettings } from 'foremanReact/Root/Context/ForemanContext';
import SelectAllAlert from './SelectAllAlert';
import { columns as defaultColumns } from './InsightsTableConstants';
import TableEmptyState from '../../../common/table/EmptyState';
import { modifySelectedRows, getSortColumnIndex } from './InsightsTableHelpers';
import Pagination from './Pagination';
import './table.scss';

const InsightsTable = ({
  page,
  perPage: urlPerPage,
  status,
  sortBy,
  sortOrder,
  hits,
  query,
  fetchInsights,
  onTableSort,
  onTableSelect,
  selectedIds,
  showSelectAllAlert,
  selectAll,
  clearAllSelection,
  error,
  isAllSelected,
  hideHost,
}) => {
  const { perPage: appPerPage } = useForemanSettings();
  const perPage = urlPerPage || appPerPage;
  const [rows, setRows] = React.useState([]);
  const [columns, setColumns] = React.useState(defaultColumns);

  // acts as componentDidMount
  useEffect(() => {
    fetchInsights({ page, perPage, query, sortBy, sortOrder });
  }, []);

  useEffect(() => {
    setRows(
      modifySelectedRows(hits, selectedIds, showSelectAllAlert, hideHost)
    );

    if (hideHost) setColumns(cols => cols.filter(c => c.id !== 'hostname'));
  }, [hits, selectedIds, hideHost]);

  return (
    <React.Fragment>
      <SelectAllAlert
        selectedIds={selectedIds}
        showSelectAllAlert={showSelectAllAlert}
        selectAll={selectAll}
        clearAllSelection={clearAllSelection}
        isAllSelected={isAllSelected}
      />
      <Table
        className="rh-cloud-recommendations-table"
        aria-label="Recommendations Table"
        onSelect={(_event, isSelected, rowId) =>
          onTableSelect(isSelected, rowId, rows, selectedIds)
        }
        canSelectAll
        sortBy={{
          index: getSortColumnIndex(columns, sortBy),
          direction: sortOrder,
        }}
        onSort={(_event, index, direction) =>
          onTableSort(columns, index, direction)
        }
        cells={columns}
        rows={rows}
        variant="compact"
      >
        <TableHeader />
        <TableBody />
      </Table>
      <TableEmptyState status={status} error={error} />
      <Pagination variant="bottom" />
    </React.Fragment>
  );
};

InsightsTable.propTypes = {
  page: PropTypes.number,
  perPage: PropTypes.number,
  status: PropTypes.string,
  sortBy: PropTypes.string,
  sortOrder: PropTypes.string,
  hits: PropTypes.array.isRequired,
  onTableSort: PropTypes.func.isRequired,
  onTableSelect: PropTypes.func.isRequired,
  selectedIds: PropTypes.object,
  showSelectAllAlert: PropTypes.bool,
  selectAll: PropTypes.func.isRequired,
  clearAllSelection: PropTypes.func.isRequired,
  fetchInsights: PropTypes.func.isRequired,
  query: PropTypes.string,
  error: PropTypes.string,
  isAllSelected: PropTypes.bool,
  hideHost: PropTypes.bool,
};

InsightsTable.defaultProps = {
  page: 1,
  perPage: null,
  status: null,
  sortBy: '',
  sortOrder: '',
  selectedIds: {},
  showSelectAllAlert: false,
  query: '',
  error: '',
  isAllSelected: false,
  hideHost: false,
};

export default InsightsTable;
