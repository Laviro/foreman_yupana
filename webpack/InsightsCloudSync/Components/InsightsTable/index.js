import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import * as actions from './InsightsTableActions';
import InsightsTable from './InsightsTable';
import InsightsTableReducer from './InsightsTableReducer';
import {
  selectPage,
  selectPerPage,
  selectSearch,
  selectStatus,
  selectSortBy,
  selectSortOrder,
  selectHits,
  selectItemCount,
  selectSelectedIds,
  selectShowSelectAllAlert,
  selectError,
  selectIsAllSelected,
} from './InsightsTableSelectors';

// map state to props
const mapStateToProps = state => ({
  query: selectSearch(state),
  page: selectPage(state),
  perPage: selectPerPage(state),
  status: selectStatus(state),
  sortBy: selectSortBy(state),
  sortOrder: selectSortOrder(state),
  hits: selectHits(state),
  itemCount: selectItemCount(state),
  selectedIds: selectSelectedIds(state),
  showSelectAllAlert: selectShowSelectAllAlert(state),
  error: selectError(state),
  isAllSelected: selectIsAllSelected(state),
});

// map action dispatchers to props
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

// export reducers
export const reducers = { table: InsightsTableReducer };

// export connected component
export default connect(mapStateToProps, mapDispatchToProps)(InsightsTable);
