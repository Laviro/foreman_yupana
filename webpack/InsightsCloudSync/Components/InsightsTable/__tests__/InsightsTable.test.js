import { testComponentSnapshotsWithFixtures } from '@theforeman/test';

import InsightsTable from '../InsightsTable';
import { tableProps } from './fixtures';

const fixtures = {
  'render with Props': tableProps,
};

describe('InsightsTable', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(InsightsTable, fixtures));
});
