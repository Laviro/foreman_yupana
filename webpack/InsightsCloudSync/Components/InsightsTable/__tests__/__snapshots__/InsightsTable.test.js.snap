// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InsightsTable rendering render with Props 1`] = `
<Fragment>
  <SelectAllAlert
    clearAllSelection={[Function]}
    isAllSelected={false}
    selectAll={[Function]}
    selectedIds={Object {}}
    showSelectAllAlert={false}
  />
  <Table
    aria-label="Recommendations Table"
    borders={true}
    canSelectAll={true}
    canSortFavorites={true}
    cells={
      Array [
        Object {
          "sortKey": "hostname",
          "title": "Hostname",
          "transforms": Array [
            [Function],
            [Function],
          ],
        },
        Object {
          "sortKey": "title",
          "title": "Recommendation",
          "transforms": Array [
            [Function],
            [Function],
          ],
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "sortKey": "total_risk",
          "title": "Total risk",
          "transforms": Array [
            [Function],
            [Function],
          ],
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "title": "Playbook",
          "transforms": Array [
            [Function],
          ],
        },
      ]
    }
    className="recommendations-table"
    contentId="expanded-content"
    dropdownDirection="down"
    dropdownPosition="right"
    expandId="expandable-toggle"
    gridBreakPoint="grid-md"
    isStickyHeader={false}
    onSelect={[Function]}
    onSort={[Function]}
    ouiaSafe={true}
    role="grid"
    rowLabeledBy="simple-node"
    rows={Array []}
    selectVariant="checkbox"
    sortBy={
      Object {
        "direction": "desc",
        "index": 3,
      }
    }
    variant="compact"
  >
    <TableHeader />
    <Component />
  </Table>
  <TableEmptyState
    error={null}
    status="RESOLVED"
  />
  <Pagination
    variant="bottom"
  />
</Fragment>
`;
